### PR TITLE
Add RT-Thread RTOS Support

### PR DESCRIPTION
**Pull request by wuhanstudio**
Hi,

This PR adds support for [RT-Thread](https://www.rt-thread.io/), which is a RTOS that supports [150+ boadrs](https://www.rt-thread.io/board.html).

Previously, I ported [libsodium ](https://packages.rt-thread.org/en/search.html?search=libsodium) to [RT-Thread Package center](https://packages.rt-thread.org/en/index.html), but recently I notice that libhydrogen may fit better for embedded devices.

Hope you are interested,
Han

ezgif com-gif-maker

### CODE REVIEW

**Comment by jedisct1** [comment]
Thanks!

Does the `Sconscript` file need to be there, though?

`scons` is a generic build system, but the configuration seems specific to RT-Thread.

**Comment by wuhanstudio** [comment]
Hi, 

The `SConscritpt` file is used by the package manager, so that we can add packages directly from the [package center](https://packages.rt-thread.org/en/index.html) using [RT-Thread Studio](https://www.rt-thread.io/studio.html), then the IDE and build system will automatically add `include path` and `source files`  to the project.

Indeed, RT-Thread uses `scons` as its build system. Some of the configurations are specific to RT-Thread because the `SConscript` is used in conjunction with `KConfig`, so that we can also use `menuconfig` to automatically config and download packages, similar to `linux` and `buildroot`

https://github.com/RT-Thread/packages/blob/master/security/libsodium/Kconfig

https://user-images.githubusercontent.com/15157070/167189495-47fbed8d-6611-4049-8092-8bc5d3930f31.mp4

**Comment by jedisct1** [comment]
Could the Sconscript file live in a target-specific directory such as `./rt-thread/` ?

**Comment by wuhanstudio** [comment]
> ./rt-thread/

Sure, similar to [lvgl](https://github.com/lvgl/lvgl), I can move RT-Thread specific configurations into the folder `./rt-thread/` .

But a general `SConscript`  still need to be in the root directory that points to RTT specific configurations in `./rt-thread/`.

```
# RT-Thread building script

import os
from building import *

objs = []
cwd  = GetCurrentDir()

objs = objs + SConscript(cwd + './rt-thread/SConscript')

Return('objs')
```

**Comment by jedisct1** [comment]
Mmmm.. that doesn't really help. Typing `scons` here is going to fail everywhere besides RT-Thread.

**Comment by wuhanstudio** [comment]
> Mmmm.. that doesn't really help. Typing `scons` here is going to fail everywhere besides RT-Thread.

The `scons` build system does not conflic with CMake or Makefile. It fails because the program cannot be compiled alone without the source code of kernels, drivers and etc.

A Quick Start for simulation in QEMU (RT-Thread & libsodium):

```
# Ubuntu 20.04
sudo apt install scons gcc-arm-none-eabi qemu-system-arm
git clone https://github.com/rt-thread/rt-thread
cd rt-thread/bsp/qemu-vexpress-a9/
scons --menuconfig 
# Choose RT-Thread online packages --> security packages --> libsodium --> select examples and tests
# Exit and Save
source ~/.env/env.sh
pkgs --update
# Add "-Wno-deprecated-declarations -Wno-unused-variable" into packages/libsodium/SConscript (fixed in the latest PR)
scons
./qemu-nographic.sh
# Ctrl + A, then X to exit
```

https://user-images.githubusercontent.com/15157070/167261609-eaf89175-d37c-4d51-a6bd-e6e895afa67b.mp4

File: SConscript [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Import('RTT_ROOT')
+Import('rtconfig')
+from building import *
+
+objs = []
+cwd     = GetCurrentDir()
+
+src     = Glob('hydrogen.c')
```
Comment (jedisct1) [inline SConscript:8]: For a single file, globbing is probably not required :)

File: impl/random/rtthread.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
+// Please enable RT_USING_HWCRYPTO and RT_HWCRYPTO_USING_RNG
```
Comment (jedisct1) [inline impl/random/rtthread.h:1]: What does that mean? Is it possible to check for that in the scons file and abort/explain how to enable it if they aren't enabled?

**Comment by jedisct1** [comment]
It remains unfortunate that the file has to be in the root directory.

When used on a different target, would it be possible to fail with a message explaining that the scons script only supports RT-Thread, and that Makefiles/zig/cmake should be used instead?

File: impl/random/rtthread.h (hunk 2 above)
Comment (wuhanstudio) [reply impl/random/rtthread.h:1]: > What does that mean? Is it possible to check for that in the scons file and abort/explain how to enable it if they aren't enabled?

Actually these two options are automatically enabled once the pacakge libhygrogen is selected.

Thus, this it just a friently reminder that libhydrogen depends on hardware PRNG, or this line can be removed.

File: SConscript (hunk 1 above)
Comment (wuhanstudio) [reply SConscript:8]: Yes, I'll update this.

File: impl/random/rtthread.h (hunk 2 above)
Comment (wuhanstudio) [reply impl/random/rtthread.h]: This line hase be removed, because these options will automatically be enabled by the package manager.

**Comment by wuhanstudio** [comment]
> It remains unfortunate that the file has to be in the root directory.
> 
> When used on a different target, would it be possible to fail with a message explaining that the scons script only supports RT-Thread, and that Makefiles/zig/cmake should be used instead?

Indeed, it will be much better to print a message telling developers `scons` should be used with RT-Thread.

I've updated the SConscript to be like this:

```
from building import *

if GetDepend('PKG_USING_LIBHYDROGEN'):
    Import('RTT_ROOT')
    Import('rtconfig')

    objs    = []
    cwd     = GetCurrentDir()

    src     = 'hydrogen.c'
    path    = [cwd]

    group   = DefineGroup('libhydrogen', src, depend = ['PKG_USING_LIBHYDROGEN'], CPPPATH = path)

    Return('group')

else:
    print('The scons script only supports RT-Thread, please use Makefiles/zig/cmake.') 
```

**Pull request closed without merging**